File: 67.二进制求和.cpp
/*
 * @lc app=leetcode.cn id=67 lang=cpp
 *
 * [67] 二进制求和
 */

// @lc code=start
class Solution
{
public:
    string addBinary(string a, string b)
    {
        int len = max(a.size(), b.size());
        int u = 0;
        reverse(a.begin(), a.end());
        reverse(b.begin(), b.end());
        for (int i = 0; i < len; i++)
        {
            u += i < a.size() ? (a.at(i) == '1') : 0;
            u += i < b.size() ? (b.at(i) == '1') : 0;
            ans.push_back((u % 2) ? '1' : '0');
            u /= 2;
        }
        if (u)
        {
            ans.push_back('1');
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }

private:
    string ans;
};
// @lc code=end

File: JavaScript/67.二进制求和.js
/*
 * @lc app=leetcode.cn id=67 lang=javascript
 *
 * [67] 二进制求和
 */

// @lc code=start
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function (a, b) {
  let carry = 0; // 进位标志
  let result = ""; // 结果字符串

  // 从后往前遍历两个二进制字符串
  let i = a.length - 1;
  let j = b.length - 1;

  while (i >= 0 || j >= 0 || carry > 0) {
    // 取出当前位置的数字，若已遍历完则视为0
    const digitA = i >= 0 ? parseInt(a[i]) : 0;
    const digitB = j >= 0 ? parseInt(b[j]) : 0;
    const sum = digitA + digitB + carry;
    carry = Math.floor(sum / 2);
    result = (sum % 2) + result;
    i--;
    j--;
  }

  return result;
};
// @lc code=end
